File: C++/Problems/Problems/problem_451.cpp
#include "problem_451.h"

#include <algorithm>
#include <unordered_map>

bool problem_451::test() {
    std::string s = "Aabb";

    std::string expected = "bbAa";

    auto result = frequencySort(s);

    std::cout << result << std::endl;

    return expected == result;
}

std::string problem_451::frequencySort(std::string s) {
    std::unordered_map<char, int> counts;
    for (auto c : s)
    {
        if (counts.contains(c))
            counts[c]++;
        else
            counts[c] = 1;
    }

    int chars = counts.size();
    std::pair<int, char> pairs[chars];
    int i = 0;
    for (auto count : counts)
    {
        pairs[i] = std::pair(-count.second, count.first);
        i++;
    }
    std::sort(pairs, pairs + chars);
    std::string result = "";
    for (auto pair : pairs)
    {
        result +=  std::string(-pair.first, pair.second);
    }

    return result;
}

File: Python/Problems/Problems/problem451.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set
from collections import Counter

from .problem import Problem


class Problem451(Problem):
    def test(self) -> bool:
        s = "Aabb"

        expected = "bbAa"

        result = self.frequencySort(s)

        print(result)

        return expected == result

    def frequencySort(self, s: str) -> str:
        counts = dict()
        for c in s:
            if c in counts:
                counts[c] += 1
            else:
                counts[c] = 1
        pairs = []
        for count in counts.items():
            pairs.append((count[1], count[0]))
        pairs.sort(key=lambda pair: (-pair[0], pair[1]))
        result = ''
        for pair in pairs:
            result += pair[1] * pair[0]

        return result